### PR TITLE
handle dates like 13 September 2024 18:29:58 +0000

### PR DESCRIPTION
**Pull request by mayoms**
Closes # .

## Changes proposed in this pull request

Updates date parsing to support invalid dates like `13 September 2024 18:29:58 +0000`

### CODE REVIEW

**Comment by bcardarella** [comment]
I think we've always just assumed 8601 for the datetime format but looking over RFC 2822's date time specification (https://www.rfc-editor.org/rfc/rfc2822#section-3.3) I don't see any mention of 8601.

@mayoms what is generating this datetime format?

**Comment by mayoms** [comment]
@bcardarella it's from an email we received, but failed to parse

**Comment by bcardarella** [comment]
Does the raw email happen to say what the client is?

**Comment by mayoms** [comment]
It does not (no user-agent or otherwise identifying header that I can see) -- the bad date format was in the "received" header if that helps. All others are properly formatted.

**Comment by bcardarella** [comment]
Gotcha. In that case I don't think the request in #170 makes sense, at least in the sense that we shouldn't patch this library to allow for greater compatibility of datetime formats.

@andrewtimberlake thoughts?

**Comment by mayoms** [comment]
> Gotcha. In that case I don't think the request in https://github.com/DockYard/elixir-mail/issues/170 makes sense, at least in the sense that we shouldn't patch this library to allow for greater compatibility of datetime formats.

Sorry I'm not sure I follow this statement? Are you saying this PR negates the need for more user control? Or, are you saying this PR doesn't make sense in light of #170 ?

**Comment by bcardarella** [comment]
if someone encounters a date that should be accepted but isn't parsing I'd prefer for them to file a bug so the library can be fixed than providing API such as in #170. That opens the door to "why not an override for this" and "why not an override for that". That's going to be a nightmare to maintain.

**Comment by mayoms** [comment]
FWIW as a user, I like this idea better than the custom parser I suggested.

**Comment by mayoms** [comment]
Any additional thoughts here? If you'd like (and are decided on the direction) I could add the error tuple code to this PR -- lmk.

**Comment by bcardarella** [comment]
@andrewtimberlake your call, I'm 👍

**Comment by andrewtimberlake** [comment]
Yes, thanks @mayoms, add the error tuple to this PR

File: lib/mail/parsers/rfc_2822.ex [hunk 1]
```diff
@@ -44,7 +59,12 @@ defmodule Mail.Parsers.RFC2822 do
   Parses a RFC2822 timestamp to a DateTime with timezone
 
   [RFC2822 3.3 - Date and Time Specification](https://tools.ietf.org/html/rfc2822#section-3.3)
+
+  Also supports obsolete format described in [RFC2822 4.3](https://datatracker.ietf.org/doc/html/rfc2822#section-4.3)
+  and invalid timestamps encountered in the wild. The return value will be either a UTC DateTime, or an error tuple
+  returning the invalid date string.
```
Comment (mayoms) [inline lib/mail/parsers/rfc_2822.ex:65]: I'm very willing to take any feedback or omit this altogether, but I thought it'd make sense to document the change in return value. I'd considered either listing the formats supported or suggesting checking out `Mail.Parsers.RFC2822Test` to see which 'invalid formats' are supported, but wanted to get some feedback on whether or not y'all wanted specific documentation for non compliant date support.

Comment (bcardarella) [reply lib/mail/parsers/rfc_2822.ex:65]: I think the test suite is sufficient for the supported formats. We should aim to be as permissive as any email client  so I'm sure there will be other instances of datetime formats we aren't supported that we need to adapt to

File: test/mail/parsers/rfc_2822_test.exs [hunk 2]
```diff
@@ -252,6 +252,35 @@ defmodule Mail.Parsers.RFC2822Test do
...
+          "0#{idx}"
+        else
+          idx
+        end
+
+      {:ok, datetime, 0} = DateTime.from_iso8601("2024-#{idx}-13 18:29:58Z")
+      assert to_datetime("13 #{long_month} 2024 18:29:58 +0000") == datetime
+    end)
+
+    assert to_datetime("invalid date string") == {:error, "invalid date string"}
```
Comment (andrewtimberlake) [inline test/mail/parsers/rfc_2822_test.exs:283]: I suggest pulling this out to it’s own test to highlight the behaviour of unexpected date strings apart from known invalid date strings.

**Comment by andrewtimberlake** [comment]
Thanks for this @mayoms

**Pull request merged**